x-pack/plugins/security_solution/public/common/components/threat_match/index.tsx

```diff
@@ -32,135 +29,81 @@ const MyButtonsContainer = styled(EuiFlexItem)`
-  entriesToDelete: ThreatMapEntries[];
-}
-
 interface ThreatMatchComponentProps {
-  listItems: ThreatMapEntries[];
+  mappingEntries: ThreatMapEntries[];
   indexPatterns: DataViewBase;
   threatIndexPatterns: DataViewBase;
-  onChange: (arg: OnChangeProps) => void;
+  'id-aria'?: string;
+  'data-test-subj'?: string;
+  onMappingEntriesChange: (newValue: ThreatMapEntries[]) => void;
 }
 
 export const ThreatMatchComponent = ({
-  listItems,
+  mappingEntries,
   indexPatterns,
   threatIndexPatterns,
-  onChange,
+  'id-aria': idAria,
+  'data-test-subj': dataTestSubj,
+  onMappingEntriesChange,
 }: ThreatMatchComponentProps) => {
-  const [{ entries, entriesToDelete, andLogicIncluded }, dispatch] = useReducer(reducer(), {
-    ...initialState,
-  });
-
-  const setUpdateEntries = useCallback(
-    (items: ThreatMapEntries[]): void => {
-      dispatch({
-        type: 'setEntries',
-        entries: items,
-      });
-    },
-    [dispatch]
-  );
-
-  const setDefaultEntries = useCallback(
-    (item: ThreatMapEntries): void => {
-      dispatch({
-        type: 'setDefault',
-        initialState,
-        lastEntry: item,
-      });
-    },
-    [dispatch]
-  );
-
   const handleEntryItemChange = useCallback(
     (item: ThreatMapEntries, index: number): void => {
-      const updatedEntries = [
-        ...entries.slice(0, index),
-        {
-          ...item,
-        },
-        ...entries.slice(index + 1),
-      ];
-
-      setUpdateEntries(updatedEntries);
+      const updatedEntries = mappingEntries.slice();
+
+      updatedEntries.splice(index, 1, item);
+
+      onMappingEntriesChange(updatedEntries);
     },
-    [setUpdateEntries, entries]
+    [mappingEntries, onMappingEntriesChange]
   );
 
   const handleDeleteEntryItem = useCallback(
-    (item: ThreatMapEntries, itemIndex: number): void => {
+    (item: ThreatMapEntries, index: number): void => {
       if (item.entries.length === 0) {
-        const updatedEntries = [...entries.slice(0, itemIndex), ...entries.slice(itemIndex + 1)];
-        // if it's the only item left, don't delete it just add a default entry to it
-        if (updatedEntries.length === 0) {
-          setDefaultEntries(item);
-        } else {
-          setUpdateEntries([...entries.slice(0, itemIndex), ...entries.slice(itemIndex + 1)]);
-        }
+        const updatedEntries = mappingEntries.slice();
+
+        updatedEntries.splice(index, 1);
+
+        onMappingEntriesChange(updatedEntries);
       } else {
-        handleEntryItemChange(item, itemIndex);
+        handleEntryItemChange(item, index);
       }
     },
-    [handleEntryItemChange, setUpdateEntries, entries, setDefaultEntries]
+    [mappingEntries, onMappingEntriesChange, handleEntryItemChange]
   );
 
-  const handleAddNewEntryItemEntry = useCallback((): void => {
-    const lastEntry = entries[entries.length - 1];
-    const { entries: innerEntries } = lastEntry;
+  const handleOrClick = useCallback((): void => {
```

Comment:
I included a bullet item `ThreatMatchComponent was refactored to be a controlled component to simplify Threat Match Mapping editable component` though it doesn't contain details. I'll update the description.

The main point of refactoring `ThreatMatchComponent` has internal state which doesn't work well when we wanna have external state. I had bugs related to state restoring. The refactoring removed internal state from `ThreatMatchComponent` making it a controlled component.

In case you need to refresh what controlled/uncontrolled component mean you could look through [legacy React docs page](https://legacy.reactjs.org/docs/uncontrolled-components.html) (new docs don't seem to contain the same page).